monads/src/FailCont/Main.hs

```diff
@@ -10,10 +10,25 @@ data Error
   deriving (Show, Eq)
 
 addInts :: String -> String -> FailCont r Error Int
-addInts = undefined 
+addInts x y = case (readMaybe x, readMaybe y) of
+  (Nothing, _)       -> FailCont $ \_ fail -> fail EmptyInput
+  (_, Nothing)       -> FailCont $ \_ fail -> fail (ParseFailed y)
+  (Just a, Just b)   -> FailCont $ \success _ -> success (a + b)
 
-divInts :: String -> String -> FailCont r Error Int 
-divInts = undefined 
+divInts :: String -> String -> FailCont r Error Int
+divInts x y = case (readMaybe x, readMaybe y) of
```

Comment:
Create a function that does `readMaybe` and returns `Either` -- this will make your code much more readable.